library/std/src/os/uefi/env.rs

```diff
@@ -0,0 +1,130 @@
+//! UEFI-specific extensions to the primitives in `std::env` module
+
+use super::raw::{BootServices, RuntimeServices, SystemTable};
+use crate::ffi::c_void;
+use crate::io;
+use crate::mem::MaybeUninit;
+use crate::ptr::NonNull;
+use crate::sync::atomic::{AtomicPtr, Ordering};
+use r_efi::efi::{Guid, Handle};
+use r_efi::system;
+
+static GLOBAL_SYSTEM_TABLE: AtomicPtr<SystemTable> = AtomicPtr::new(crate::ptr::null_mut());
+static GLOBAL_SYSTEM_HANDLE: AtomicPtr<c_void> = AtomicPtr::new(crate::ptr::null_mut());
+
+/// Initializes Global Atomic Pointers to SystemTable and Handle.
+/// Should only be called once in the program execution under normal circumstances.
+/// The caller should ensure that the pointers are valid.
+pub(crate) fn init_globals(handle: NonNull<c_void>, system_table: NonNull<SystemTable>) {
+    GLOBAL_SYSTEM_TABLE.store(system_table.as_ptr(), Ordering::SeqCst);
+    GLOBAL_SYSTEM_HANDLE.store(handle.as_ptr(), Ordering::SeqCst);
+}
+
+/// Get the SystemTable Pointer.
+#[unstable(feature = "uefi_std", issue = "100499")]
+pub fn get_system_table() -> Option<NonNull<SystemTable>> {
+    NonNull::new(GLOBAL_SYSTEM_TABLE.load(Ordering::SeqCst))
+}
+
+/// Get the SystemHandle Pointer.
+#[unstable(feature = "uefi_std", issue = "100499")]
+pub fn get_system_handle() -> Option<NonNull<c_void>> {
+    NonNull::new(GLOBAL_SYSTEM_HANDLE.load(Ordering::SeqCst))
+}
+
+/// Get the BootServices Pointer.
+#[unstable(feature = "uefi_std", issue = "100499")]
+pub fn get_boot_services() -> Option<NonNull<BootServices>> {
+    let system_table = get_system_table()?;
+    let boot_services = unsafe { (*system_table.as_ptr()).boot_services };
+    NonNull::new(boot_services)
+}
+
+/// Get the RuntimeServices Pointer.
+#[unstable(feature = "uefi_std", issue = "100499")]
+pub fn get_runtime_services() -> Option<NonNull<RuntimeServices>> {
+    let system_table = get_system_table()?;
+    let runtime_services = unsafe { (*system_table.as_ptr()).runtime_services };
+    NonNull::new(runtime_services)
+}
+
+/// Open Protocol on a handle
+/// Implemented using `EFI_BOOT_SERVICES.OpenProtocol()`
```

Comment:
Well, `OpenProtocol` has been part of UEFI since EFI 1.1 which was released in 2005. So I'm not sure how useful it will be (since this would need changes in upstream `r-efi` to make `open_protocol` an Option).
Also the specification states the following:

> The HandleProtocol() function is still available for use by old EFI applications and drivers. However,
all new applications and drivers should use EFI_BOOT_SERVICES.OpenProtocol() in place of
HandleProtocol().

Reply:
After some discussion with my mentors, it seems to find a device that doesn't at least support `OpenProtocol` would be almost impossible. Therefore, I am resolving this conversation. Feel free to open it again if you have anything more to add.